File: js/tree-iter.js
function inOrderTraversalNonRecursive(root) {
    if (!root) {
        console.log("🌱 Tree is empty. Nothing to traverse!");
        return;
    }
    var stack = [];
    var traversalResult = [];
    console.log("🚀 Starting in-order traversal");
    // Push root node with moment 1
    stack.push({ node: root, moment: 1 });
    console.log("\uD83D\uDCE5 PUSH: Node ".concat(root.value, ", Moment 1"));
    while (stack.length > 0) {
        console.log("\n\uD83D\uDCCA STACK: ".concat(stack.map(function (item) { return "(".concat(item.node.value, ",").concat(item.moment, ")"); }).join(', ')));
        var _a = stack.pop(), node = _a.node, moment = _a.moment;
        console.log("\uD83D\uDCE4 POP: Node ".concat(node.value, ", Moment ").concat(moment));
        if (moment === 1) {
            // Push the same node with moment 2
            stack.push({ node: node, moment: 2 });
            console.log("\uD83D\uDCE5 PUSH: Node ".concat(node.value, ", Moment 2"));
            // If there's a left child, push it with moment 1
            if (node.left) {
                stack.push({ node: node.left, moment: 1 });
                console.log("\uD83D\uDCE5 PUSH: Node ".concat(node.left.value, ", Moment 1"));
            }
            else {
                console.log("\uD83C\uDF43 No left child for node ".concat(node.value));
            }
        }
        else if (moment === 2) {
            // Visit the node in moment 2 (in-order traversal)
            traversalResult.push(node.value);
            console.log("\uD83C\uDFAF VISIT: Node ".concat(node.value));
            // If there's a right child, push it with moment 1
            if (node.right) {
                stack.push({ node: node.right, moment: 1 });
                console.log("\uD83D\uDCE5 PUSH: Node ".concat(node.right.value, ", Moment 1"));
            }
            else {
                console.log("\uD83C\uDF43 No right child for node ".concat(node.value));
            }
        }
        // No action for moment 3 in in-order traversal
    }
    console.log("\n\u2705 In-order traversal complete!");
    console.log("\uD83D\uDD22 Result: ".concat(traversalResult.join(', ')));
}
// Example usage
var tree = {
    value: 10,
    left: {
        value: 8,
        left: null,
        right: {
            value: 12,
            left: null,
            right: null
        }
    },
    right: {
        value: 7,
        left: null,
        right: null
    }
};
inOrderTraversalNonRecursive(tree);

File: haskell/tree-iter.hs
import Control.Monad.State
import Control.Monad.Writer
import Data.List (intercalate)

data Tree a = Empty | Node a (Tree a) (Tree a) deriving (Show)
type Stack a = [(Tree a, Int)]
type Log = [String]

preOrderWithLogging :: Show a => Tree a -> (Log, [a])
preOrderWithLogging tree = runWriter $ 
    evalStateT (go []) [(tree, 1)]
  where
    go :: Show a => [a] -> StateT (Stack a) (Writer Log) [a]
    go acc = do
      stack <- get
      tell ["\n** Current Stack: " ++ showStack stack]
      tell ["\n** Current Result: " ++ show acc]
      
      case stack of
        [] -> do
          tell ["\n** Stack empty! Traversal complete."]
          return (reverse acc)
        
        (Empty, _):rest -> do
          tell ["\n** Popped Empty node, skipping"]
          put rest
          go acc
        
        (Node val left right, 1):rest -> do
          tell ["\n** Moment 1: Visiting node " ++ show val]
          tell ["\n   - Adding value to result"]
          tell ["\n   - Pushing same node with moment 2"]
          put ((Node val left right, 2):rest)
          
          when (left /= Empty) $ do
            tell ["\n   - Pushing left child with moment 1"]
            modify (\s -> (left, 1):s)
          
          tell ["\n   - Updated stack: " ++ showStack ((Node val left right, 2):rest)]
          go (val:acc)
        
        (Node val _ right, 2):rest -> do
          tell ["\n** Moment 2: Processing node " ++ show val]
          tell ["\n   - Pushing same node with moment 3"]
          put ((Node val Empty right, 3):rest)
          
          when (right /= Empty) $ do
            tell ["\n   - Pushing right child with moment 1"]
            modify (\s -> (right, 1):s)
            
          stack' <- get
          tell ["\n   - Updated stack: " ++ showStack stack']
          go acc
        
        (Node val _ _, 3):rest -> do
          tell ["\n** Moment 3: Finishing node " ++ show val]
          tell ["\n   - Nothing to do in pre-order, just removing from stack"]
          put rest
          go acc

showStack :: Show a => Stack a -> String
showStack stack = intercalate ", " 
                 $ map (\(node, moment) -> 
                         case node of 
                           Empty -> "(Empty," ++ show moment ++ ")"
                           Node v _ _ -> "(" ++ show v ++ "," ++ show moment ++ ")") 
                 $ stack

-- Example usage:
exampleTree :: Tree Int
exampleTree = Node 10 
                (Node 8 
                  (Node 2 Empty Empty) 
                  (Node 12 Empty Empty))
                (Node 7 Empty Empty)

runExample :: IO ()
runExample = do
  let (logs, result) = preOrderWithLogging exampleTree
  putStrLn "Pre-order traversal with logging:"
  mapM_ putStrLn logs
  putStrLn $ "\nFinal result: " ++ show result